opal/util/info.c

```diff
@@ -165,95 +191,95 @@ int opal_info_dup_mode (opal_info_t *info, opal_info_t **newinfo,
 {
     int err, flag;
     opal_info_entry_t *iterator;
-    char savedkey[OPAL_MAX_INFO_KEY + 1]; // iterator->ie_key has this as its size
-    char savedval[OPAL_MAX_INFO_VAL];
-    char *valptr, *pkey;
+
+    const char *pkey;
     int is_IN_key;
     int exists_IN_key, exists_reg_key;
 
     OPAL_THREAD_LOCK(info->i_lock);
     OPAL_LIST_FOREACH(iterator, &info->super, opal_info_entry_t) {
-// If we see an __IN_<key> key but no <key>, decide what to do based on mode.
-// If we see an __IN_<key> and a <key>, skip since it'll be handled when
-// we process <key>.
-         is_IN_key = 0;
-         exists_IN_key = 0;
-         exists_reg_key = 0;
-         pkey = iterator->ie_key;
-         if (0 == strncmp(iterator->ie_key, OPAL_INFO_SAVE_PREFIX,
-             strlen(OPAL_INFO_SAVE_PREFIX)))
+        // If we see an __IN_<key> key but no <key>, decide what to do based on mode.
+        // If we see an __IN_<key> and a <key>, skip since it'll be handled when
```

Comment:
I hate this whole "prefix" thing, but I realize it's not of this PR's making.

...but now that we have proper and info and string classes, can we do something better?

This is a whole hunk of new/heavily modified code here.  Do we have any tests that test what the whole "prefix" thing was doing to know if this new code is working properly?  (i.e., preserving user-set values, and exercising the different cases)

Reply:
My understanding for having this mechanism is that we allow info subscribers to change the value and are not allowed to hand out a different  value in `MPI_Info_get`. I don't know why this is useful, components that don't support an info value should just ignore it. 

@markalle Can you shed some light on why we allow subscribers to change the value? What was the use case for that?

Reply:
I have dim recollections that this has something to do with distinguishing -- and preserving -- the info values that users set vs. the info values that Open MPI set.  I don't remember the exact semantics, but it may have had something to do with what MPI_INFO_GET is supposed to return (or the info that results from MPI_COMM_DUP and the like?) -- i.e., what values the user set vs. the values that Open MPI is actually using.

Reply:
I don't think this actually works correctly on `master` then: `MPI_Info_get` will only query the key provided by the user and not check whether there is a backup key prefixed with `__IN_`. So if we set another value for a key we will actually hand out that new value to the user (my reading of the standard is that this is incorrect behavior). The only place where I can see the prefixed key being used is in `opal_info_dup_mode` which is called from `MPI_Info_dup`. There it is used to drop new values and use the backup instead. 

Allowing subscriber callbacks to alter an info value is problematic for another reason: assuming we have more than one component subscribing to an info value on the same object (which I don't think we have right now), the value they see depends on the order in which they subscribed (the one set by the user or the one changed by any previous component). That doesn't seem right to me...

Reply:
> I don't think this actually works correctly on `master` then: `MPI_Info_get` will only query the key provided by the user and not check whether there is a backup key prefixed with `__IN_`. So if we set another value for a key we will actually hand out that new value to the user (my reading of the standard is that this is incorrect behavior). The only place where I can see the prefixed key being used is in `opal_info_dup_mode` which is called from `MPI_Info_dup`. There it is used to drop new values and use the backup instead.

It's quite possible that I'm not remembering correctly.

I think that the intent was to preserve 2 sets of values for cases of conflict... but I don't remember what the precise conflict was -- something to do with what OMPI set vs. what the user set.  I don't remember exactly why we had to maintain both values.  ☹️  Maybe it had something to do with the imprecise description of INFO_GET (which has since been fixed for MPI-4)...?  Something to do with that weird phrasing about "MPI returns the value actually being used", or somesuch...?

> Allowing subscriber callbacks to alter an info value is problematic for another reason: assuming we have more than one component subscribing to an info value on the same object (which I don't think we have right now), the value they see depends on the order in which they subscribed (the one set by the user or the one changed by any previous component). That doesn't seem right to me...

Agreed.

Reply:
@bosilca Pointed me to §12.2.7 (`MPI_Win_set_info` and `MPI_Win_get_info`) which says that the info object retrieved through `MPI_Win_get_info` should contain the actual values used for hints, not necessarily the ones set through `MPI_Win_set_info`. So it makes sense that the info subscriber can signal the use of a different value.

However, I still don't understand why we store a backup then. The info object returned by `MPI_Win_get_info` is a new info object that is duplicated from the info object stored in the window object. It is completely independent of the info object used during window creation and if I `MPI_Win_dup` the info returned by `MPI_Win_get_info` I still want to see these values, not some backup. @markalle What am I missing here?

Reply:
Agreed -- we need to figure out what the right behavior is for dup'ing infos, etc.

That being said, I still kinda hate the whole `_IN` prefix thing.  Can we just have storage for 2 strings -- one that the user set and one that Open MPI is using?  In many (most?) cases, they will be the same (and could be ref counted, or one of them could even be NULL or something).  But more importantly, it would make the code more clear about the explicit choice to save a 2nd value sometimes.

Reply:
I was hoping to figure out whether we actually need this functionality at all. If not, we can remove it and don't have to bother fixing it...

Reply:
> I was hoping to figure out whether we actually need this functionality at all. If not, we can remove it and don't have to bother fixing it...

Sorry: I wasn't disagreeing. All I was trying to say was: _if_ we need the "have 2 set possible values" functionality, then it would be better to ditch the `_IN` prefix stuff and just explicitly store 2 strings (e.g., in two different members/variables).